src/compiler/crystal/compiler.cr

```diff
@@ -296,6 +298,19 @@ module Crystal
       result
     end
 
+    private def with_file_lock(output_dir)
+      {% if flag?(:win32) %}
+        # TODO: use flock when it's supported in Windows
+        yield
+      {% else %}
+        File.open(File.join(output_dir, "compiler.lock"), "w") do |file|
+          file.flock_exclusive do
+            yield
+          end
+        end
+      {% end %}
```

Comment:
Shouldn't lockfile be deleted (cleaned up) afterwards? IMO that would be a good practice even though it's not strictly necessary.

```suggestion
      {% end %}
    ensure
      File.delete(File.join(output_dir, "compiler.lock"))
```

Reply:
I'm afraid that's another cause for a race condition and I'd like to avoid that

Reply:
`... rescue nil` should "fix" those ;) IMO that's cleaner than a stray lockfile.

Reply:
nobody sees that directory

Reply:
I wonder... can you flock a directory? Because that would be the simplest thing to do. But I can't try it now.